### ISSUE TITLE
[Typescript pack] support biome for linter and formatter

### ISSUE THREAD

**Issue by dannd4** [AstroNvim/astrocommunity#962]
### Describe the new feature

Enable biome lsp for lspconfig and null-ls when project has @biomejs/biome dependency or has biome.json

**Comment by Uzaaft** [comment]
Great idea! Feel free to open up a PR. :)

**Comment by willparsons** [comment]
I'm running into this problem at the moment with biome and eslint+prettier bumping into each other. I'll try and sort out the config on my end and then contribute it back here.

**Comment by willparsons** [comment]
This is a bit annoying as the 95% of use cases is that you're either in a project that uses biome OR eslint+prettier, which is fine to configure, but it's possible that some people may want something like eslint with biome for formatting or some other configuration.

There's even the argument about HOW we run the LSP. I'd say we expect it to be in the projects dependencies and invoke it via `npx/yarn/pnpm biome lsp-proxy`, but they COULD use mason...

Do you guys want to cover all use cases or just go for the most common one?

**Comment by ammuench** [comment]
I'm in two projects currently where we're using biome as the formatter and eslint as the linter while migrating a bunch of stuff.

Could we check for the existence of a `biome.json` config and use that to make decisions? It might be short-sighted but in my head any project where biome has `"formatter": { "enabled": true }` turned on I think would be the true formatter even if prettier is still in node modules somewhere, since you have to manually opt-into Biome formatting by declaring it.  Same thing in reverse for `"linter": { "enabled": true }` (thought I can't think of many scenarios where you'd use biome as a linter and prettier as a formatter)

**Comment by Uzaaft** [comment]
> This is a bit annoying as the 95% of use cases is that you're either in a project that uses biome OR eslint+prettier, which is fine to configure, but it's possible that some people may want something like eslint with biome for formatting or some other configuration.
> 
> There's even the argument about HOW we run the LSP. I'd say we expect it to be in the projects dependencies and invoke it via `npx/yarn/pnpm biome lsp-proxy`, but they COULD use mason...
> 
> Do you guys want to cover all use cases or just go for the most common one?

Go for the most common one, and install it trough mason for now.

**Comment by willparsons** [comment]
I've started work on this here: https://github.com/AstroNvim/astrocommunity/pull/1006/files

Having everything supported is pretty finicky :/

**Comment by taskylizard** [comment]
That PR was closed, so I'll like to continue on this, but before that I'd like to investigate what approach we can take to suit everyone's needs.

**Comment by Uzaaft** [comment]
@taskylizard  Sure. I'd say copy the approach done with the ruff pack would be a great starting point. @willparsons  What was the bits that were finicky?

**Comment by taskylizard** [comment]
Sure thing, it's been a long while since I last contributed here, so it will take me a bit. I'll look around the different packs and see what solution we can apply.

**Comment by Uzaaft** [comment]
Great!

**Comment by taskylizard** [comment]
I decided to kang @willparsons's work on their PR and improve on bit, seems I'm a bit lost on conform support and the first task on the PR todo. I'll send a draft PR soon though.

**Comment by willparsons** [comment]
The awkward bits were just deciding when to switch biome on and off, when it should take precedency over eslint or prettier in a repo with both, how do you determine if it's 'configured' to run, etc. Not to mention doing it for null-ls, conform and nvim-lint.

**Comment by Uzaaft** [comment]
@taskylizard I like that idea.  I believe prettierd does that internally(i.e nodemodules vs the bundled option)
As for 2: Is that a field in the biome json config?

**Comment by taskylizard** [comment]
Nope, it's something we will provide from the pack, if the user wants to use a downloaded biome binary or something.

**Comment by Uzaaft** [comment]
How will you do that? Is that an opt that will be passed to the pack?

**Comment by Uzaaft** [comment]
The preferred way to install biome is to have it in your project. I think we should do that and only that for now.

**Comment by willparsons** [comment]
Yeah biome is prone to changing lints on version changes (not major version), so if your biome lsp differs from the installed version you will get 2 different results that can conflict.

You can go around using the package manager by going straight into the node_modules and running the binary.

**Comment by azdanov** [comment]
Would it make sense, then, to refactor the `typescript` pack into separate ones?

- TypeScript
- ESLint
- Prettier
- Biome.js

Otherwise, it will become a giant 1000+ line mess if someone wants to add https://oxc.rs/ or any new tool that comes out.
